### DISCUSSION TITLE
6th normal form

### DISCUSSION THREAD

**Discussion by thiemo-taken** [pgmodeler/pgmodeler#1875] [Ideas]
Hi

Are there any plans to support the generation of code in the 6th normal form, like the [anchor modelling](https://www.anchormodeling.com/) does?

Kind regards

Thiemo

**Comment by rkhaotix** [comment]
Hi @thiemo-taken 

6th normal form requires decomposing tables, and currently pgModeler does not change the structure of tables already designed by the users (unless when using some specific relationships). 

Anyway, I can think on how to add such feature to the tool in the future.

**Reply by thiemo-taken** [reply]
Hi @rkhaotix (nice nick, any connection to Asterix by chance?)

Thanks for having a thought on it.

I personally would not change the modelling at all. If you have seen the graphics of the anchor modelling tool, exploded, it is a nightmare in dots and lines. Completely useless to get an overview and totally unnecessary, in my eyes. I was just thinking of taking the 3rd NF model and generate 6th NF code of it. I.e. instead of populating a table with all its attributes defined, make two+ tables, one (anchor) with the business key (unique key) and for each other attribute a separate table (with foreign key to its anchor table). All the payload attributes should be not-null, probably all attributes.

**Comment by rkhaotix** [comment]
> Hi @rkhaotix (nice nick, any connection to Asterix by chance?)

No haha, despite I'm being a great fan of Asterix! :)
This nick is kinda a homage to one of my preferred games: Knuckles Chaotix (Sega 32x) :D

> I personally would not change the modelling at all. If you have seen the graphics of the anchor modelling tool, exploded, it is a nightmare in dots and lines. Completely useless to get an overview and totally unnecessary, in my eyes. 

Yeah, it's awful!

> I was just thinking of taking the 3rd NF model and generate 6th NF code of it. I.e. instead of populating a table with all its attributes defined, make two+ tables, one (anchor) with the business key (unique key) and for each other attribute a separate table (with foreign key to its anchor table). All the payload attributes should be not-null, probably all attributes.

That's the problem with pgModeler. The model itself must be in 6th normal form to create a code in 6th normal form.
The better approach, at first glance, is to provide a feature that converts the model or a set of tables into a desired normal form and then generates the code to reflect the current normal form of the model/tables.

**Comment by thiemo-taken** [comment]
> No haha, despite I'm being a great fan of Asterix! :) This nick is kinda a homage to one of my preferred games: Knuckles Chaotix (Sega 32x) :D

Love those comix ;-) since my childhood.

> > I was just thinking of taking the 3rd NF model and generate 6th NF code of it. I.e. instead of populating a table with all its attributes defined, make two+ tables, one (anchor) with the business key (unique key) and for each other attribute a separate table (with foreign key to its anchor table). All the payload attributes should be not-null, probably all attributes.
> 
> That's the problem with pgModeler. The model itself must be in 6th normal form to create a code in 6th normal form. The better approach, at first glance, is to provide a feature that converts the model or a set of tables into a desired normal form and then generates the code to reflect the current normal form of the model/tables.

Ok, I was rather thinking of another code generator and not the existing one. But model conversion would be fine by me. Out of interest. Is somewhere an xsd for the dbm files?

**Comment by rkhaotix** [comment]
> Out of interest. Is somewhere an xsd for the dbm files?

pgModeler database model files are ruled/validated by DTD files. You can check them here:
https://github.com/pgmodeler/pgmodeler/tree/develop/assets/schemas/xml/dtd

The DTD of the entire database model file is defined here:
https://github.com/pgmodeler/pgmodeler/blob/develop/assets/schemas/xml/dtd/dbmodel.dtd

**Comment by thiemo-taken** [comment]
> pgModeler database model files are ruled/validated by DTD files. You can check them here: https://github.com/pgmodeler/pgmodeler/tree/develop/assets/schemas/xml/dtd
> 
> The DTD of the entire database model file is defined here: https://github.com/pgmodeler/pgmodeler/blob/develop/assets/schemas/xml/dtd/dbmodel.dtd

I am a bit puzzled. Are the types not defined. At least for constraints, it seems to me, that the value is just free text.

**Reply by rkhaotix** [reply]
Which types do you mean? PostgreSQL data types?

If so, they are hard-coded here:
https://github.com/pgmodeler/pgmodeler/blob/develop/libs/libcore/src/pgsqltypes/pgsqltype.cpp

**Comment by thiemo-taken** [comment]
> Which types do you mean? PostgreSQL data types?
> 
> If so, they are hard-coded here: https://github.com/pgmodeler/pgmodeler/blob/develop/libs/libcore/src/pgsqltypes/pgsqltype.cpp

No, I meant constraint types, but I suppose, it works the same. And I reckon that having the types defined in the dtd too, might be overkill.

Hm, just noticed that there must be two lists of constraint types. The one in https://github.com/pgmodeler/pgmodeler/blob/develop/libs/libcore/src/pgsqltypes/constrainttype.cpp looks like the GUI, but the values in the dbm file are different.

**Comment by thiemo-taken** [comment]
In installed QtCreator to fiddle around a bit, but already fail with the inclusion of xmlparser.h.
